### PR TITLE
Check for CPLEX in the PATH environment variable

### PR DESCRIPTION
**Pull request by dourouc05**
This should ease installation in some cases.

### CODE REVIEW

**Comment by dourouc05** [comment]
I used the "new" functions `Sys.isapple()` and the like: as CPLEX.jl does not support Julia 0.7- (the `Project.toml` file seems to imply that), maybe it's time to clean up the build file to remove anything from before Julia 1.0?

File: deps/build.jl [hunk 1]
```diff
@@ -9,46 +9,66 @@ end
...
-        print(f,"const libcplex = ")
+        print(f, "const libcplex = ")
         show(f, path) # print with backslashes excaped on windows
         println(f)
     end
 end
 
-@static if (VERSION >= v"0.7.0-DEV.3382" && Sys.isapple()) || (VERSION < v"0.7.0-DEV.3382" && is_apple())
-    Libdl.dlopen("libstdc++",Libdl.RTLD_GLOBAL)
+@static if Sys.isapple()
```
Comment (blegat) [inline deps/build.jl]: Thanks for the cleanup!

File: deps/build.jl [hunk 2]
```diff
@@ -9,46 +9,66 @@ end
...
 
 base_env = "CPLEX_STUDIO_BINARIES"
 
+cplex_path = try
+        @static if Sys.isapple() || Sys.isunix()
+            read(`which cplex`, String)
+        elseif Sys.iswindows()
+            read(`where cplex`, String)
+        end
+    catch
```
Comment (blegat) [inline deps/build.jl]: The indendation of this block should be decreased by 1 IMO as `try` is in an unindented block

File: deps/build.jl [hunk 3]
```diff
@@ -9,46 +9,66 @@ end
...
+        nothing
+    end
+if cplex_path !== nothing
+    cplex_path = splitdir(strip(cplex_path))[1]
```
Comment (blegat) [inline deps/build.jl]: Could you add a comment on usually what's the value of `cplex_path` after and before this line ?

File: deps/build.jl (hunk 1 above)
Comment (dourouc05) [reply deps/build.jl]: Shall I make a PR with similar changes throughout the package? (The question being: is there a rewrite effort in progress, like for Gurobi and GLPK?)

File: deps/build.jl [hunk 4]
```diff
@@ -2,53 +2,75 @@
...
+        read(`which cplex`, String)
+    elseif Sys.iswindows()
+        read(`where cplex`, String)
+    end
+catch
+    nothing
+end
+if cplex_path !== nothing
+    # Extract the path to the folder containing the CPLEX executable.
+    cplex_path = splitdir(strip(cplex_path))[1]
```
Comment (blegat) [inline deps/build.jl]: Isn't `splitdir(s)[1]` equivalent to `dirname(s)` ?

File: deps/build.jl [hunk 5]
```diff
@@ -2,53 +2,75 @@
...
+                    push!(libnames, joinpath(d, "cplex$(v)0"))
                 else
-                    push!(libnames,joinpath(d,"cplex$(v)"))
+                    push!(libnames, joinpath(d, "cplex$(v)"))
                 end
             end
         end
+        if cplex_path !== nothing
+            push!(libnames, joinpath(cplex_path, "cplex$(v)"))
+            push!(libnames, joinpath(cplex_path, "cplex$(v)0"))
```
Comment (blegat) [inline deps/build.jl]: Why adding the `0` ? It is already added in the `wincpxvers` array, isn't it ?

Comment (dourouc05) [reply deps/build.jl]: Actually, I duplicated what is done above, when checking with the environment variable.

Comment (blegat) [reply deps/build.jl]: Yes, but to be sure that zeros are in the `wincpxvers` array, I would do
```julia
base_wincpxvers = ["128", "129"]
wincpxvers = [base_wincpxvers; base_wincpxvers .* "0"]
```
and same for `cpxvers` for consistency

Comment (dourouc05) [reply deps/build.jl]: I just did this. How about having only one array instead of two (wincpxvers and cpxvers)?

Comment (blegat) [reply deps/build.jl]: No idea why there are two, they used to be different but I don't know whether we might need two different list again in the future.

**Comment by dourouc05** [comment]
As this PR is approved, could it get merged?

File: deps/build.jl [hunk 6]
```diff
@@ -2,53 +2,73 @@
...
         push!(libnames, "libcplex$v.dylib")
-    elseif (VERSION >= v"0.7.0-DEV.3382" && Sys.isunix()) || (VERSION < v"0.7.0-DEV.3382" && is_unix())
+    elseif Sys.isunix()
         push!(libnames, "libcplex$v.so")
         if haskey(ENV, base_env)
             push!(libnames, joinpath(ENV[base_env], "libcplex$v.so"))
         end
+        if cplex_path !== nothing
+            push!(libnames, joinpath(cplex_path, "libcplex$v.so"))
+        end
```
Comment (odow) [inline deps/build.jl]: Why only add this is `isiunix`? What about the `is apple` and the `.dylib`?

File: deps/build.jl [hunk 7]
```diff
@@ -2,53 +2,73 @@
...
     end
 end
 
-const wincpxvers = ["128", "1280", "129", "1290"]
-@static if (VERSION >= v"0.7.0-DEV.3382" && Sys.iswindows()) || (VERSION < v"0.7.0-DEV.3382" && is_windows())
+base_wincpxvers = ["128", "129"]
+wincpxvers = [base_wincpxvers; base_wincpxvers .* "0"]
```
[the remark this answers is not in the document]
Comment (dourouc05) [reply deps/build.jl]: That question has already been raised. I suppose your comment means that one of them can leave.

File: deps/build.jl (hunk 6 above)
Comment (dourouc05) [reply deps/build.jl]: I did something in the latest commit, but I have no possibility to test on macOS… Does that look right to you? (Maybe I should merge the two `@static if`?)

Comment (odow) [reply deps/build.jl]: Yeah, can't this whole file just be tidied up and simplified?

**Comment by dourouc05** [comment]
I've just simplified the code quite a bit by merging all code paths for the different OSes. 

I really like what you did for Gurobi, the diagnoses. Should a similar implementation go into this PR or rather another one? 

This also points to a discrepancy between the two packages: for CPLEX.jl, just finding the CPLEX executable on the path is enough; for Gurobi.jl, the environment variable must be set. I think that using the version on the path is a very reasonable default (and should be implemented in Gurobi.jl). To be sure that the Julia package finds the version the user really wants, maybe we should print it when building the Julia package?

**Comment by dourouc05** [comment]
I just pushed a few more changes, mostly cleaning things up. I also tried to extract the version within `build.jl`, which could also be a nice sanity check, but it's a nightmare to make it work, as you cannot just include the relevant functions from the rest of the package (I start getting "world age" errors). Here is what I was writing (the `eval`s avoid errors about not giving a Ptr, but a Tuple{String, String}, even though that's what all CPLEX.jl is doing… I guess this is a Julia bug); I was stuck at typing issues for `stat` (), but I'm giving up right now: 

```
function get_version(path)
    # Cannot use CPLEX.version
    version = "nothing"

    if Sys.iswindows()
        # Create environment.
        stat = Vector{Cint}(undef, 1)
        env = eval(:(ccall(("CPXopenCPLEX", path), stdcall, Ptr{Cvoid}, (Ptr{Cint},), stat)))

        # Get version.
        version_char_ptr = eval(:(ccall(("version", path), stdcall, Ptr{Cchar}, (Ptr{Cvoid},), env)))
        if version_char_ptr == C_NULL
            error("CPLEX: error getting version")
        end
        version = unsafe_string(version_char_ptr)

        # Close CPLEX.
        eval(:(ccall(("CPXcloseCPLEX", path), stdcall, Cint, (Ptr{Cvoid},), env)))
    end

    return version
end
```

This works locally (Windows 10 1709, Julia 1.2 RC2): 

```
(v1.2) pkg> build CPLEX
  Building CPLEX → `C:\Users\XXX\.julia\dev\CPLEX\deps\build.log`
 Resolving package versions...
```

**Comment by odow** [comment]
You could add a print with `Using CPLEX found in location $(libname)` if you like.

**Comment by dourouc05** [comment]
I just did that in the last commit. The only problem is that this is printed out in the log file, but I'm not sure anything else is possible.

**Pull request merged**